runtime/lua/vim/treesitter/query.lua

```diff
@@ -740,84 +793,52 @@ function M.list_predicates()
   return vim.tbl_keys(predicate_handlers)
 end
 
-local function xor(x, y)
-  return (x or y) and not (x and y)
-end
-
-local function is_directive(name)
-  return string.sub(name, -1) == '!'
-end
-
 ---@private
+---@param preds ProcessedPredicate[]
 ---@param match TSQueryMatch
 ---@param source integer|string
-function Query:match_preds(match, source)
-  local _, pattern = match:info()
-  local preds = self.info.patterns[pattern]
-
-  if not preds then
-    return true
-  end
-
-  local captures = match:captures()
-
-  for _, pred in pairs(preds) do
+function Query:match_preds(preds, pattern_i, captures, source)
+  for _, pred in ipairs(preds) do
     -- Here we only want to return if a predicate DOES NOT match, and
     -- continue on the other case. This way unknown predicates will not be considered,
     -- which allows some testing and easier user extensibility (#12173).
-    -- Also, tree-sitter strips the leading # from predicates for us.
-    local is_not = false
-
-    -- Skip over directives... they will get processed after all the predicates.
-    if not is_directive(pred[1]) then
-      local pred_name = pred[1]
-      if pred_name:match('^not%-') then
-        pred_name = pred_name:sub(5)
-        is_not = true
-      end
 
-      local handler = predicate_handlers[pred_name]
+    local processed_name = pred[1]
+    local should_match = pred[2]
+    local pattern = pred[3]
```

Comment:
Maybe can just change this to
```suggestion
    local pred_name = pred[3][1]
```
and then use no handler for `pred_name` below? Since this is the only place `pattern` is used currently. Then maybe we can also change the parameter name from `pattern_i` to `pattern` which seems clearer to me

Reply:
also passed to the handler as the last argument.

Regarding the rename, `pattern_i` is clearer since it's not the actual pattern, but its index.

Reply:
Ah thanks, that makes sense. nevermind then